### PR TITLE
EIP-7549 gRPC (part 1)

### PR DESCRIPTION
**Pull request by rkapka**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

Implements the following gRPC endpoints:
- `ListAttestationsElectra`
- `ListIndexedAttestationsElectra`
- `AttestationPoolElectra`
- `SubmitAttesterSlashingElectra`

Logic shared by pre and post-Electra functions are extracted to generic helper functions.

### CODE REVIEW

File: beacon-chain/rpc/prysm/v1alpha1/beacon/attestations.go [hunk 1]
```diff
@@ -67,45 +69,107 @@ func (bs *Server) ListAttestations(
 			return nil, status.Errorf(codes.Internal, "Could not fetch attestations: %v", err)
 		}
 	case *ethpb.ListAttestationsRequest_Epoch:
+		if q.Epoch >= params.BeaconConfig().ElectraForkEpoch {
```
Comment (james-prysm) [inline beacon-chain/rpc/prysm/v1alpha1/beacon/attestations.go:72]: this seems kinda hacky, is this a change in the spec?

File: beacon-chain/rpc/prysm/v1alpha1/beacon/attestations.go [hunk 2]
```diff
@@ -136,29 +207,205 @@ func (bs *Server) ListIndexedAttestations(
...
+			var expected T
+			return nil, status.Errorf(codes.Internal, "Attestation is of the wrong type (expected %T, got %T)", expected, att)
+		}
+		atts = append(atts, a)
+	}
+
+	return atts, nil
+}
+
+func blockIndexedAttestations[T ethpb.IndexedAtt](
```
Comment (james-prysm) [inline beacon-chain/rpc/prysm/v1alpha1/beacon/attestations.go:386]: nice use of generics

File: beacon-chain/rpc/prysm/v1alpha1/beacon/attestations.go [hunk 3]
```diff
@@ -67,45 +69,107 @@ func (bs *Server) ListAttestations(
...
+	}, nil
+}
+
+// ListAttestationsElectra retrieves attestations by block root, slot, or epoch.
+// Attestations are sorted by data slot by default.
+//
+// The server may return an empty list when no attestations match the given
+// filter criteria. This RPC should not return NOT_FOUND. Only one filter
+// criteria should be used.
+func (bs *Server) ListAttestationsElectra(ctx context.Context, req *ethpb.ListAttestationsRequest) (*ethpb.ListAttestationsElectraResponse, error) {
```
Comment (james-prysm) [inline beacon-chain/rpc/prysm/v1alpha1/beacon/attestations.go:120]: Most of the logic below is copied, is there a way to reduce this? or probably not

Comment (james-prysm) [reply beacon-chain/rpc/prysm/v1alpha1/beacon/attestations.go:120]: wonder if we can use the interface here or a utility function to build the response to reduce the copy paste here

File: beacon-chain/rpc/prysm/v1alpha1/beacon/attestations.go (hunk 1 above)
Comment (rkapka) [reply beacon-chain/rpc/prysm/v1alpha1/beacon/attestations.go:72]: this is gRPC, not Beacon API

File: beacon-chain/rpc/prysm/v1alpha1/beacon/attestations.go (hunk 3 above)
Comment (rkapka) [reply beacon-chain/rpc/prysm/v1alpha1/beacon/attestations.go:120]: Yes, it's a bit duplicated but the returned types are non-compatible. I would probably have to create an interface that both return types implement and given that we are getting close to deprecating gRPC, I don't think it's worth it. There's also close to 0 chance that there will be a third similar function as I don't expect attestations to change again in the F-fork.

**Review by james-prysm — APPROVED** [review]
conditionally approving but at the ACDC there was discussions on changing this EIP further need to find where the followup issue is.

**Pull request merged**